plutus-tx/src/PlutusTx/AssocMap.hs

```diff
@@ -84,17 +84,17 @@ instance (FromData k, FromData v) => FromData (Map k v) where
           traverseFromBuiltin = go
             where
                 go :: BI.BuiltinList (BI.BuiltinPair BI.BuiltinData BI.BuiltinData) -> Maybe [(k, v)]
-                go l = BI.chooseList l (const (pure [])) (\_ -> let tup = BI.head l in liftA2 (:) (liftA2 (,) (fromBuiltinData $ BI.fst tup) (fromBuiltinData $ BI.snd tup)) (go (BI.tail l))) ()
+                go l = BI.chooseList l (const (pure [])) (\_ -> let ~tup = BI.head l in liftA2 (:) (liftA2 (,) (fromBuiltinData $ BI.fst tup) (fromBuiltinData $ BI.snd tup)) (go (BI.tail l))) ()
 
 instance (UnsafeFromData k, UnsafeFromData v) => UnsafeFromData (Map k v) where
-    unsafeFromBuiltinData d = let es = BI.unsafeDataAsMap d in Map $ mapFromBuiltin es
+    unsafeFromBuiltinData d = let ~es = BI.unsafeDataAsMap d in Map $ mapFromBuiltin es
```

Comment:
> so surely this isn't helping?

Wrong - nonstrict bindings are more likely to be inlined since they are effect safe. This reduces the budgets for several test cases.

Reply:
That doesn't make sense: this is used exactly once, and it's always going to be used because it's immediately the argument to a function. I don't doubt it changes the budget but it sounds like we don't understand why. We should find out and add a comment.

Reply:
> I don't doubt it changes the budget but it sounds like we don't understand why.

If `es` is strict, it isn't effectSafe to inline, is it? If it's not inlined, then it's more expensive.

Reply:
Extremely marginally more expensive! Literally just the overhead of the let-binding. Also, I would expect `firstEffectfulTerm` to handle this anyway. Regardless, it doesn't sound like we know what's actually happening!

Reply:
@michaelpj I checked this, and it is exactly because of inlining. The only difference of the two PIRs is:

```
\(d : data) -> go (unMapData d)
```

vs.

```
\(d : data) ->	
  let	
    !es : list (pair data data) = unMapData d	
  in	
  go es
```

The budget difference is indeed marginal - 0.6%. As to why `firstEffectfulTerm` didn't handle this: the first effectful term here is `go`, not `es`. We can potentially make it smarter by having it look at the definition of `go` and realize that `go` is a lambda, but that's a separate conversation.

Reply:
Okay, then this needs a comment saying "remove this when we fix that stuff". Otherwise it's a mystery and I'm just going to forget and remove it later as clearly unnecessary.

Reply:
Added
> -- The `~` here enables `BI.unsafeDataAsMap d` to be inlined, which reduces costs slightly.
  -- Without the `~`, the inliner would consider it not effect safe to inline.
  -- The other `~`s in this module are there for the same reason.
  -- We can remove all the `~`s once we make the inliner smart enough to inline them.